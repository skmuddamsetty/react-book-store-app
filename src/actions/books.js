// import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_BOOK
export const addBook = book => ({
  type: 'ADD_BOOK',
  book
});
// REMOVE_BOOK
export const removeBook = ({ id } = {}) => ({
  type: 'REMOVE_BOOK',
  id
});
// EDIT_BOOK

export const editBook = (id, updates) => ({
  type: 'EDIT_BOOK',
  id,
  updates
});

export const startAddBook = (bookData = {}) => {
  return dispatch => {
    const { description = '', note = '', price = 0, createdAt = 0 } = bookData;
    const book = { description, note, price, createdAt };
    return database
      .ref('books')
      .push(book)
      .then(ref => {
        dispatch(addBook({ id: ref.key, ...book }));
      })
      .catch();
  };
};
