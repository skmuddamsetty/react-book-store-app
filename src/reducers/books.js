// Books reducer
const booksReducerDefaultState = [];

export const booksReducer = (state = booksReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_BOOK':
      //   return state.concat(action.book);
      return [...state, action.book];
    case 'REMOVE_BOOK':
      return state.filter(book => book.id !== action.id);
    case 'EDIT_BOOK':
      return state.map(book => {
        if (book.id === action.id) {
          return { ...book, ...action.updates };
        } else {
          return book;
        }
      });
    default:
      return state;
  }
};
