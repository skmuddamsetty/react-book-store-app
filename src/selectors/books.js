// get Visible Books
export const getVisibleBooks = (
  books,
  { text, sortBy, startDate, endDate }
) => {
  return books
    .filter(book => {
      const startDateMatch =
        typeof startDate !== 'number' || book.createdAt >= startDate;
      const endDateMatch =
        typeof endDate !== 'number' || books.createdAt <= endDate;
      const textMatch = book.description
        .toLowerCase()
        .includes(text.toLowerCase());
      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === 'price') {
        return a.price < b.price ? -1 : 1;
      } else {
        return 0;
      }
    });
};
