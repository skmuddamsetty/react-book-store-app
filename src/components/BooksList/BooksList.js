import React from 'react';
import { connect } from 'react-redux';
import BooksListItem from './BooksListItem/BooksListItem';
import { getVisibleBooks } from '../../selectors/books';

export const BooksList = props => (
  <div>
    {
      props.books === undefined || props.books.length === 0 ?
        (<p>No Books</p>) : (props.books.map((book, idx) => {
          return <BooksListItem {...book} key={idx} />;
        }))
    }
  </div>
);

const mapStateToProps = state => {
  return {
    books: getVisibleBooks(state.books, state.filters)
  };
};
export default connect(mapStateToProps)(BooksList);
