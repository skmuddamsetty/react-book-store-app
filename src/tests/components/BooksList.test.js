import React from 'react';
import { shallow } from 'enzyme';
import { BooksList } from '../../components/BooksList/BooksList';
import { books } from '../fixtures/books';
import toJSON from 'enzyme-to-json';

test('should render books list with books', () => {
    const wrapper = shallow(<BooksList books={books} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
})

test('should render books list with empty message', () => {
    const wrapper = shallow(<BooksList books={[]} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
})