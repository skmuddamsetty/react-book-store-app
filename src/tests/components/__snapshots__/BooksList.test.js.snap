// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render books list with books 1`] = `
<div>
  <BooksListItem
    createdAt={0}
    description="React Book"
    id="1"
    key="0"
    note="React Book Notes"
    price={25}
  />
  <BooksListItem
    createdAt={1}
    description="Angular Book"
    id="2"
    key="1"
    note="Angular Book Notes"
    price={10}
  />
  <BooksListItem
    createdAt={2}
    description="Node Book"
    id="3"
    key="2"
    note="Node Book Notes"
    price={20}
  />
</div>
`;

exports[`should render books list with empty message 1`] = `
<div>
  <p>
    No Books
  </p>
</div>
`;
